Network/DNS/API/Resolv.hs

```diff
@@ -0,0 +1,152 @@
+            Right l  -> tryUntilSuccess
+                            (\a -> contactDNSResolverAt (show a) mport mto mr)
+                            (pureDns $ errorDns $ errorMsg "non of the IPv4 address worked for this domain " ++ (show $ toBytes fqdn))
+                            l
+            Left err -> pureDns $ errorDns $ errorMsg $ "error while lookupA: " ++ show err
+
+-- | Attempt to create a resolv seed which contact the given hostname
+contactDNSResolverAt :: HostName         -- ^ an ip address (IPv6 or IPv4)
+                     -> Maybe PortNumber -- ^ port number
+                     -> Maybe Seconds    -- ^ timeout
+                     -> Maybe Int        -- ^ retry
+                     -> DnsIO DNS.ResolvSeed
+contactDNSResolverAt hostname mport mto mr =
+    catchAny
+         (DNS.makeResolvSeed $ resolvConf { DNS.resolvInfo = resolvInfo })
+         (\ex -> pureDns $ errorDns $ errorMsg $ "could not create a resolv seed from the received IP: " ++ show hostname ++ " error is: " ++ show ex)
+  where
+    errorMsg :: String -> String
+    errorMsg str = "Network.DNS.API.Client.contactDNSResovlerAt: " ++ str
+
+    resolvInfo :: DNS.FileOrNumericHost
+    resolvInfo = maybe (DNS.RCHostName hostname) (\p -> DNS.RCHostPort hostname p) mport
+    resolvConf :: DNS.ResolvConf
+    resolvConf = let r1 = DNS.defaultResolvConf
+                     r2 = maybe r1 (\(Seconds to) -> r1 { DNS.resolvTimeout = (fromIntegral to) * 3000 * 3000 }) mto
+                 in maybe r2 (\r  -> r2 { DNS.resolvRetry = r }) mr
+
+-- | Create the ResolvSeed using the given parameters
+--
+-- Will first attempt to create a ResolvSeed with the HostName information
+-- (The DNS Server address and the port) if provided.
+-- We will try any IPv4 or IPv6 available from the given domain name
+--
+-- If still not found, we will then fall back to the default DNS Resolver
+-- (/etc/resolv.conf in linux/mac).
+--
+-- If it still fails, then fall back to a more generic DNS configuration by contacting NS Resolver at address 8.8.8.8
+makeResolvSeedSafe :: Maybe ByteString -- ^ the DNS Server to contact (Domain Name, IPv4 or IPv6)
+                   -> Maybe PortNumber -- ^ port number (useless without the DNS Server to contact)
+                   -> Maybe Seconds    -- ^ timeout
+                   -> Maybe Int        -- ^ retry
+                   -> DnsIO DNS.ResolvSeed
+makeResolvSeedSafe mfqdn mport mto mr = do
+    attemptToContactAGlobalResolver <|> attemptToContactTheDefaultResolvers
+  where
+    attemptToContactTheDefaultResolvers :: DnsIO DNS.ResolvSeed
+    attemptToContactTheDefaultResolvers =
+        getFirstResolvSeed
+            [ resolvConf
+            , resolvConf { DNS.resolvInfo = DNS.RCHostName "8.8.8.8" }
+            ]
+
+    attemptToContactAGlobalResolver :: DnsIO DNS.ResolvSeed
+    attemptToContactAGlobalResolver =
+        maybe (pureDns $ errorDns $ errorMsg "no Domain Name provided, fall back to the local resolv configuration")
```

Comment:
hmmm, you are right, I just copied/pasted it from the previous code. I'll change it too.